cpp/src/arrow/filesystem/azurefs_test.cc

```diff
@@ -154,6 +137,328 @@ TEST(AzureFileSystem, OptionsCompare) {
+  std::shared_ptr<io::InputStream> stream;
+  ASSERT_OK_AND_ASSIGN(stream, fs_->OpenInputStream(PreexistingObjectPath()));
+
+  std::array<char, 1024> buffer{};
+  std::int64_t size;
+  ASSERT_OK_AND_ASSIGN(size, stream->Read(buffer.size(), buffer.data()));
+
+  EXPECT_EQ(std::string(buffer.data(), size), kLoremIpsum);
+}
+
+TEST_F(TestAzureFileSystem, OpenInputStreamStringBuffers) {
+  std::shared_ptr<io::InputStream> stream;
+  ASSERT_OK_AND_ASSIGN(stream, fs_->OpenInputStream(PreexistingObjectPath()));
+
+  std::string contents;
+  std::shared_ptr<Buffer> buffer;
+  do {
+    ASSERT_OK_AND_ASSIGN(buffer, stream->Read(16));
+    contents.append(buffer->ToString());
+  } while (buffer && buffer->size() != 0);
+
+  EXPECT_EQ(contents, kLoremIpsum);
+}
+
+TEST_F(TestAzureFileSystem, OpenInputStreamInfo) {
+  // TODO: When implemented use ASSERT_OK_AND_ASSIGN(info,
+  // fs->GetFileInfo(PreexistingObjectPath()));
+  arrow::fs::FileInfo info(PreexistingObjectPath(), FileType::File);
+
+  std::shared_ptr<io::InputStream> stream;
+  ASSERT_OK_AND_ASSIGN(stream, fs_->OpenInputStream(info));
+
+  std::array<char, 1024> buffer{};
+  std::int64_t size;
+  ASSERT_OK_AND_ASSIGN(size, stream->Read(buffer.size(), buffer.data()));
+
+  EXPECT_EQ(std::string(buffer.data(), size), kLoremIpsum);
+}
+
+TEST_F(TestAzureFileSystem, OpenInputStreamEmpty) {
+  const auto path_to_file = "empty-object.txt";
+  const auto path = PreexistingContainerPath() + path_to_file;
+  service_client_->GetBlobContainerClient(PreexistingContainerName())
+      .GetBlockBlobClient(path_to_file)
+      .UploadFrom(nullptr, 0);
+
+  ASSERT_OK_AND_ASSIGN(auto stream, fs_->OpenInputStream(path));
+  std::array<char, 1024> buffer{};
+  std::int64_t size;
+  ASSERT_OK_AND_ASSIGN(size, stream->Read(buffer.size(), buffer.data()));
+  EXPECT_EQ(size, 0);
+}
+
+TEST_F(TestAzureFileSystem, OpenInputStreamNotFound) {
+  ASSERT_RAISES(IOError, fs_->OpenInputStream(NotFoundObjectPath()));
+}
+
+TEST_F(TestAzureFileSystem, OpenInputStreamInfoInvalid) {
+  // TODO: When implemented use ASSERT_OK_AND_ASSIGN(info,
+  // fs->GetFileInfo(PreexistingBucketPath()));
+  arrow::fs::FileInfo info(PreexistingContainerPath(), FileType::Directory);
+  ASSERT_RAISES(IOError, fs_->OpenInputStream(info));
+
+  // TODO: When implemented use ASSERT_OK_AND_ASSIGN(info,
+  // fs->GetFileInfo(NotFoundObjectPath()));
+  arrow::fs::FileInfo info2(PreexistingContainerPath(), FileType::NotFound);
+  ASSERT_RAISES(IOError, fs_->OpenInputStream(info2));
+}
+
+TEST_F(TestAzureFileSystem, OpenInputStreamUri) {
+  ASSERT_RAISES(Invalid, fs_->OpenInputStream("abfss://" + PreexistingObjectPath()));
+}
+
+TEST_F(TestAzureFileSystem, OpenInputStreamReadMetadata) {
+  const std::string object_name = "OpenInputStreamMetadataTest/simple.txt";
+
+  service_client_->GetBlobContainerClient(PreexistingContainerName())
+      .GetBlobClient(PreexistingObjectName())
+      .SetMetadata(Azure::Storage::Metadata{{"key0", "value0"}});
```

Comment:
The implementation of `ReadMetadata()` from https://github.com/apache/arrow/pull/12914 returns only user defined metadata. I've left that implementation as it was for now and written this test but to be honest I don't think this is correct. I think really we want to return the [`BlobProperties`](https://github.com/Azure/azure-sdk-for-cpp/blob/1e013e5fe90d742b9b0ef2210e1cca1e638c18ec/sdk/storage/azure-storage-blobs/inc/azure/storage/blobs/rest_client.hpp#L1761) which includes far more than justuser defined metadata. Looking at the [GCS version](https://github.com/Tom-Newton/arrow/blob/08e5206a69ec1b5e8bb13fad0e14c22510cdb877/cpp/src/arrow/filesystem/gcsfs_test.cc#L1073-L1133) it looks like it gets similar information to [`BlobProperties`](https://github.com/Azure/azure-sdk-for-cpp/blob/1e013e5fe90d742b9b0ef2210e1cca1e638c18ec/sdk/storage/azure-storage-blobs/inc/azure/storage/blobs/rest_client.hpp#L1761).

If we agree that returning [`BlobProperties`](https://github.com/Azure/azure-sdk-for-cpp/blob/1e013e5fe90d742b9b0ef2210e1cca1e638c18ec/sdk/storage/azure-storage-blobs/inc/azure/storage/blobs/rest_client.hpp#L1761) would be better I can create a github issue for it.